backend/src/main/java/com/javahelp/backend/domain/user/register/UserRegisterInteractor.java

```diff
@@ -0,0 +1,126 @@
+import com.javahelp.model.user.UserPassword;
+
+/**
+ * An interactor to make new {@link User} and {@link UserPassword} for all types of users
+ */
+public abstract class UserRegisterInteractor <T extends IUserRegisterInputBoundary,
+        U extends UserInfo>{
+
+    IUserStore userStore;
+
+    /**
+     * Creates a new {@link UserRegisterInteractor} instance
+     *
+     * @param userStore  {@link IUserStore} to use
+     */
+    public UserRegisterInteractor(IUserStore userStore) {
+        this.userStore = userStore;
+    }
+
+    /**
+     * Generate a new pseudo {@link User}. Private because this is exclusively for a general
+     * {@link User} type.
+     *
+     * @return a new {@link User} entity
+     */
+    private User register(T boundary){
+        return new User("", this.bundleInfo(boundary), boundary.getUsername());
+    }
+
+    /**
+     * Checking for violations for registering a {@link User} account.
+     * @param boundary an instance of {@link IUserRegisterInputBoundary} to retrieve input info
+     *
+     * @return a {@link RegisterUserResponse} model that contains message of failure at registration
+     * Null if all tests passed.
+     */
+    protected RegisterUserResponse isValidRegistering(T boundary){
+        if ((IUserStore.getDefaultImplementation().readByEmail(boundary.getEmailAddress()) != null)
+                && (IUserStore.getDefaultImplementation().readByUsername(boundary.getUsername()) != null)){
+            // If both username and email address exist, then return both errors
+            return new RegisterUserResponse(null, null, "Both the given email " +
+                    "address and username are already occupied. If that is previously yours, " +
+                    "consider logging in instead?");
+        }
+        if(IUserStore.getDefaultImplementation().readByEmail(boundary.getEmailAddress()) != null){
+            return new RegisterUserResponse(null, null, "The email" +
+                    " address is already occupied. If that is previously yours, consider logging " +
+                    "in instead?");
+        }
+        if(IUserStore.getDefaultImplementation().readByUsername(boundary.getUsername()) != null){
+            return new RegisterUserResponse(null, null, "The username" +
+                    " is already occupied. If that is previously yours, consider logging in instead?");
+        }
+
+        // Nothing goes wrong
+        return null;
+    }
+
+    /**
+     * A method for future different types of {@link User} that might need to validate
+     * their own defined attributes (i.e. beyond the uniqueness check in username and email address)
+     * @param boundary <T> interface extends {@link IUserRegisterInputBoundary} to retrieve input
+     *                 info
+     *
+     * @return true if assumptions for registering that type of {@link User} holds; false otherwise
+     */
+    protected boolean isValidated(T boundary){
```

Comment:
This method could act like the `isValidRegistering` method, and return a response with an error, or else null. This way it could communicate what exactly went wrong. Additionally, this method should not call `isValidRegistering`, it should be abstract, and not define a body. This way implementing subclasses will be forced to override it and define a validation behaviour for their use cases input boundaries. Both this and `isValidRegistering` should have to return null for the use case to proceed.